### ISSUE TITLE
Implement parallelisation

### ISSUE THREAD

**Issue by ThibHlln** [unifhy-org/unifhy#6]
Before implementing parallelisation with MPI, do we need to decide on an approach to parallelisation? How many communicators? Functional and/or domain decomposition? etc.

**Comment by mattjbr123** [comment]
Starting to think about this again...

The end-goal here is to speed up UnifHy to run JULES at least as fast as standalone-JULES (the work to decompose JULES to run on UnifHy will need to be completed to test this). 

One thing that is relatively easy to do is to make the components to run in parallel, as at a given timestep they are independent of each other. Whilst this is a "quick win" a naive implementation might not lead to much speed up if one of the components is a lot slower to run than another. It might be possible to avoid this condition by further parallelising the components themselves, and then the component that takes the most walltime to run could be sped up to better match the walltimes of the other components. However this is not trivial...

There are two types of communications that we have to manage, that:
- Between the different components (as definied by the UnifHy framework)
- Within the individual components 
- - between vertical levels 
- - between horizontal grid points

It is important to remember that UnifHy provides very limited constraint on *how* the individual components compute the variables that are prescribed by UnifHy for transfer to other components. At the moment each component must have 'initialise', 'run' and 'finalise' steps, but what happens within those is largely free to the component developer to decide. Where I am going with this is that we *could* leave it up to the individual components themselves to implement some sort of parallelism in order to compute more quickly. 

Alternatively we could implement changes in UnifHy itself in the form of certain further requirements that a component must meet so that UnifHy can handle the parallelisation. I would see this as something like requiring components to split up their code into 'science that is applied over the whole grid' (and thus can be subject to domain parallelisation) and 'science that is independent across gridpoints' (and thus doesn't need to be parallelised), then for the 'science that is applied over the whole grid' what happens in an arbitrary sub-domain (a section of the overall grid that would form one 'task' that can be run in parallel) and what is needed to be passed into and out of the sub-domain. UnifHy could then take this code, determine how many sub-domains could be created to cover the whole domain, and run each sub-domain in parallel subject to the communication constraints. (All of this just for a given timestep). This parallellisation would have to be implemented by UnifHy in a specific, fixed way for each component, no variation between paralellisation implementations amongst components, but it would have the advantage of letting UnifHy handle all the parallelisation in a fixed way. I should say that I'm honestly not sure how sensible a suggestion this is and whether it would even be possible...

Either way, we are ultimately we are looking at some sort of domain parallelisation. This is fairly trivial for components that involve no lateral communication between grid points, as each 'column' or gridpoint would be independent and thus easily parallelisable, perhaps this could be seen as the 'trivial case' of the previous paragraph. 

UnifHy also has a '.route' function that allows components to hand over control of the water/nutrient routing to UnifHy. Parallelisation of this specific function, which is where the main lateral communications occur I would guess, could be a stepping-stone on the way to a fuller parallelisation as described in the above two paragraphs, and wouldn't require any changes to the existing components that have been developed. How one parallelises a routing function though I don't know...

**Comment by mattjbr123** [comment]
Some earlier discussions on parallelisation have come to light: 

[20211012_Brainstorming_on_parallelisation.pdf](https://github.com/user-attachments/files/19893887/20211012_Brainstorming_on_parallelisation.pdf)

There are some similarities in our thinking, mainly that domain parallelisation is the way to go.

I will attempt to write down my thoughts and understanding of the conclusions of this discussion:

- To run in parallel each Component needs to define the computations that take place across each subdomain or 'Parallel Element' (PE) and what (and how?) they will pass between each other.
- Currently UniFHy has one 'Exchanger' instance that handles all the transfers between Components. To run in parallel, each **PE** will have its own Exchanger. 
- At initialisation of the Model, UniFHy will split up the domains into PEs and determine which PEs of each Component need to send/receive data from which other Components, storing this information in each PE's Exchanger.
- As far as I see it, UniFHy needs to handle the exact domain decomposition to ensure each PE can map directly into/onto those of another Component that is higher/lower resolution. As it stands the grid resolutions must strictly be integer multiples of each other, and this needs to be maintained for the PEs too. 
- However the discussion document above seems to imply that Component developers are given full freedom on how they decompose their domains, and I'm not sure how that works if this similarity between the Components' grids and PEs is to be maintained. 
- Time parallelisation seems a lot easier, as at a given timestep each Component is independent of the others and can run until the next timestep it needs information from other Components, at which point it waits until that is available. MPI would handle this automatically?

**Comment by mattjbr123** [comment]
Turns out that the discussion document was based on the assumption of UniFHy having a supermesh, which didn't happen. A supermesh would be an *independent* grid that UniFHy would translate all the Components' transfers on to in order to pass them on to the others where needed. This supermesh approach would allow the Components' to have completely different grids, rather than the current requirement they be integer multiples of each other, with the same restrictions applying to Components' PEs when parallelising. 
TL;DR The requirement of the PEs being recetangular and integer multiples of each **does** apply.

Solving this problem would require developing a supermesh which would likely take a while. Better to focus getting *some* version of parallelisation running first, which can have the grid restriction (which is no different than the grid restriction that currently exists in UniFHy for the whole domains).

**Comment by mattjbr123** [comment]
![Image](https://github.com/user-attachments/assets/d97e9b9a-8233-407f-9c0b-c5fa33e8b582)

**Comment by mattjbr123** [comment]
So conceptually I think I now understand what needs to be done. The next step is to look at the code and work out an outline of how this would be achieved and what would need to be changed where.

**Comment by bnlawrence** [comment]
Grenville may have another opinion, but there are two types of pattern that important here: couplers, and frameworks.  If we have control over the code of the components, then we can consider a framework, if not, we need to think about couplers.  A coupler basically treats each component like a single exec and organises how each talks to the others. A framework is a single exec, and knows how do do the connections.   So I'd be arguing for coupling given we think everyone is bringing their own component. BUT (with a big exception) most are bringing models which do not have any horizontal transport, and so they aren't internally gridded, so we can handle the situation where there is no domain decomposition within each component, and we can do the domain decomposition ourselves. 

Ok, the big exception is JULES, but I think for the purpose of this exercise, we can make JULES be a single column model and do all the fun stuff in unifhy.  But there is still an exception to discuss with Rich: Horizontal transport in the mositure.
 All of htat is saying maybe the way to think of this is not as you have drawn it, but like this:

<img width="1266" height="1144" alt="Image" src="https://github.com/user-attachments/assets/c6e94054-527c-4d6c-8050-4a3b64f58331" />

**Comment by jmarshrossney** [comment]
Hi Brian. Interested bystander here (and Matt's RSE colleague).

> BUT (with a big exception) most are bringing models which do not have any horizontal transport

I'd love to understand more about why this is - why models with dimensions (z, t) are sufficient for hydrological modelling, or at least the particular applications in mind by HydroJULES. I get why JULES is mostly decoupled in the horizontal plane, but my understand **was** that the lack of horizontal fluxes in JULES, making in less useful for hydrology, was a key motivator for HydroJULES.

If indeed all components are 1+1 dimensional then can't parallelisation over a grid just be factored out of Unifhy entirely? The user could just be directed to whatever the most appropriate tool is for firing off a large number of independent parallel processes on their hardware. I.e., does MPI belong in this project at all?

Cheers!

**Comment by mattjbr123** [comment]
Putting together some thoughts to guide our discussion at our next in-person meet.

- We have decided the structure for any parallelisation that needs to take place, it will be based on domain decomposition: 
  - Each Component's spatial domain will be split into rectangular subdomains
  - A given subdomain in a Component must cover the same geographic area as the corresponding subdomains in the other Components, but can be a different resolution
  - For each set of subdomains covering a particular geographic area, one subdomain per Component, one instance of UniFHy will handle the transfers between the Components in the same way it does already, using it's Exchanger
  - In limited cases, horizontal transfers are required across the subdomains, and thus between different instances of UniFHy.
  - Within each subdomain of a Component, the Component is free to do whatever it likes, subject to the existing constrains UniFHy imposes, and any additional constraints we need to impose to allow parallelism to take place where this is horizontal communication
  - [Bryan's diagram](https://github.com/unifhy-org/unifhy/issues/6#issuecomment-3435716340) shows this in picture form
  - UniFHy will decide how to best split up the horizontal domain
  - The multiple instances of UniFHy form the tasks/jobs that can be run in parallel

- We have also come to the conclusion that, for now at least, the only horizontal transfers come from the OpenWater Component, with runoffs and river routing. Other Components do not have horizontal transfers. 
- Does this also apply to JULES once it has been adapted for UniFHy?

The Key Question to answer:
- What software/technology do we use to run the multiple instances of UniFHy in parallel and handle the communication between different subdomains/instances of UniFHy. MPI would seem purpose built for this, but...
- ...UniFHy is built on CF-Python. Most objects in UniFHy are a cf.Field under the hood. CF-Python has it's own internal parallelism using Dask. Does this interfere with the way we are trying to do the parallelisation? Does it make more sense/is it possible to make use of this internal parallelism to parallelise UniFHy instead of using an external tool such as MPI? 
- Some Components will wish to use UniFHy's convenience '.route' function. How does this fit into the above parallelisation design? 

- Our intention is to test any implementation first with the [RFM OpenWater Component](https://github.com/unifhy-org/unifhycontrib-rfm/blob/main/unifhycontrib/rfm/openwater.py), as this is the one that has existing horizontal transport in it, before we look at JULES

**Comment by mattjbr123** [comment]
Notes from meeting:

- Settled on [Bryan's diagram](https://github.com/unifhy-org/unifhy/issues/6#issuecomment-3435716340) as the conceptual approach to take. In the absence of horizontal transfers this is embarrassingly parallel
- Decided that MPI is a better tool to be using than Dask for handling the parallelism, because MPI gives you explicit control over the setup - what runs where and what data is sent where, whereas Dask will decide this for itself, often sub-optimally
- The horizontal transfers will be handled only through the use of UniFHy's .route function - i.e. Components written for UniFHy will have to use this .route function if they want to do horizontal transfers. At least for now
- The simplest way of implementing the parallelisation of this function in the first instance is to gather all the data from all the parallel elements/sub-domains into one process, compute the horizontal transfers across the entire domain with the routing function exclusively with this process, and scatter the computed transfers back to the parallel elements. MPI has specific commands for this that we can wrap into UniFHy
- This may prove to be inefficient, but is a first step, and avoids having to think about halos
- UniFHy will have to be extended to set up the domain decomposition based on the domains the models/Components wish to run on. 
- The ability to update the flow-directions (which are currently assumed constant throughout the run) over time is something we might need to consider for soil moisture routing, which doesn't necessarily always go downhill!
- A concern was how the regridding inbuilt into UniFHy via CF-Python and ESMF would could be parallelised, but this (I think) is resolved if we use MPI and [Bryan's diagram](https://github.com/unifhy-org/unifhy/issues/6#issuecomment-3435716340) approach, as then the regridding is done separately/independently per-parallel element already.

@bnlawrence @rich-HJ @dmlivings do add in stuff if I forgot/misremembered.
